proxylib/kafka/parser.go

```diff
@@ -0,0 +1,206 @@
+		err := kr.PortRule.Sanitize()
+		if err != nil {
+			panic("invalid Kafka rule")
+		}
+		rules = append(rules, &kr)
+	}
+	return rules
+}
+
+type KafkaParserFactory struct{}
+
+var kafkaParserFactory *KafkaParserFactory
+
+func init() {
+	log.Info("init(): Registering kafkaParserFactory")
+	RegisterParserFactory(parserName, kafkaParserFactory)
+	RegisterL7RuleParser(parserName, KafkaRuleParser)
+}
+
+type KafkaParser struct {
+	connection *Connection
+}
+
+func (pf *KafkaParserFactory) Create(connection *Connection) interface{} {
+	p := KafkaParser{connection: connection}
+	return &p
+}
+
+func (p *KafkaParser) OnData(reply bool, reader *Reader) (OpType, int) {
+	length := reader.Length()
+	if length == 0 {
+		return NOP, 0
+	}
+
+	framelength := 4          // account for the length field
+	lenbuf := make([]byte, 8) // Peek the first eight bytes
+	n, err := reader.PeekFull(lenbuf)
+	if err == nil {
+		framelength += int(binary.BigEndian.Uint32(lenbuf[:4]))
+	} else {
+		// Need more data
+		return MORE, 8 - n
+	}
+
+	if reply {
+		// Replies are always passed as-is. No need to parse them
+		// on top of the frame length and correlation ID.
+		correlationID := binary.BigEndian.Uint32(lenbuf[4:])
+		p.connection.Log(cilium.EntryType_Response,
+			&cilium.LogEntry_GenericL7{
+				GenericL7: &cilium.L7LogEntry{
+					Proto: parserName,
+					Fields: map[string]string{
+						"CorrelationID": strconv.Itoa(int(correlationID)),
+					},
+				},
+			})
+		return PASS, framelength
+	}
+
+	// Ask for more if full frame has not been received yet
+	if length < framelength {
+		// Not enough data, ask for more and try again
+		return MORE, framelength - length
+	}
+
+	var req KafkaRequest
+	req.RequestMessage, err = kafka.ReadRequest(reader)
+	if err != nil {
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to parse Kafka request; closing Kafka connection")
+		}
+		return ERROR, int(ERROR_INVALID_FRAME_TYPE)
+	}
+
+	topics := req.GetTopics()
+	// Maintain a map of all topics in the request.
+	// We should allow the request only if all topics are
+	// allowed by the list of rules.
+	req.topics = make(map[string]struct{}, len(topics))
+	for _, topic := range topics {
+		req.topics[topic] = struct{}{}
+	}
+	logEntry := &cilium.LogEntry_GenericL7{
+		GenericL7: &cilium.L7LogEntry{
+			Proto: parserName,
+			Fields: map[string]string{
+				"APIVersion":    strconv.Itoa(int(req.GetVersion())),
+				"APIKey":        kafka_api.ApiKeyToString(req.GetAPIKey()),
+				"CorrelationID": strconv.Itoa(int(req.GetCorrelationID())),
+				"Topics":        strings.Join(topics, ","),
+			},
+		},
+	}
+	if p.connection.Matches(req) {
+		p.connection.Log(cilium.EntryType_Request, logEntry)
+		return PASS, framelength
+	}
+
+	resp, err := req.CreateAuthErrorResponse()
+	if err != nil {
+		logEntry.GenericL7.Fields["status"] = strconv.Itoa(int(kafka.ErrInvalidMessage))
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to create Kafka response")
```

Comment:
Same comment as L 163.

Reply:
Yes, but I may have changed this since you reviewed.